### DISCUSSION TITLE
Modelling the expo video recorder

### DISCUSSION THREAD

**Discussion by jonathanstiansen** [statelyai/xstate#1917] [Q&A]
Hey all,

I'm running into some problems modelling our businesses video recording via XState. 

The goal: `Expocamera.startRecording` returns a promise that, once finished recording, resolves into the URI for the recording.

The same camera must have `ExpoCamera.stopRecording` must be called before the other promise resolves. 

I'm trying to figure out a "good" way of doing this.

Initially I reached for simple services, which may still be the best way of solving it, but I wanted to find a better solution.

I wasted a bunch of time on Activities, I let the console warns to describe what doesn't happen:

``` javascript
      recordVideo: async (
        context,
        _activityEvent
      ) => {
        console.log('1')
        const { uri } = await camera.recordAsync(context.options)
        console.log('2')
        context.uri = uri
        await MediaLibrary.saveToLibraryAsync(uri).then(() => {
          console.log('recordedVideo!', uri)
          return context.addVideo(uri)
        })
        console.log('3')
        return () => {
          console.log('4')
          context.cameraRef?.stopRecording()
        }
      },
    },

```

I never see any of these print. And I have no idea why, but the stopRecording seems to be getting pressed 🤷🏽 likely because of the `async` on the function? It properly transitions away but the rest of the promise doesn't resolve.

A few other options I've considered:

1. I looked at `invoked promises` but this seems to not work with the model of accepting an event "stop_recording" - however maybe I'm just not understanding.
2. I originally had a "service" call on entering the "recording" state, and could presumably add calling a different service call to stop recording on exiting. This seems like the only good way of doing it right now but it's probably from lack of understanding.

**Comment by Andarist** [comment]
> I never see any of these print. And I have no idea why, but the stopRecording seems to be getting pressed 🤷🏽 likely because of the async on the function?

`async` doesn't suppress logging so hard to tell why you don't see those. We also can't control the lifetime of the async function as it's eager so those things should be called - just without relating to the state machine's lifetime.

> I looked at invoked promises but this seems to not work with the model of accepting an event "stop_recording" - however maybe I'm just not understanding.

Not quite sure what do u mean by the model not accepting an event here.

---

I would wrap the camera into its own service and communicate through events with the parent machine that would be responsible for managing more higher-level concerns.
